src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.Analyzer.cs

```diff
@@ -0,0 +1,298 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.InvertIf
+{
+    internal abstract partial class AbstractInvertIfCodeRefactoringProvider
+    {
+        protected enum InvertIfStyle
+        {
+            // swap if and else
+            Normal,
+            // swap subsequent statements and if body
+            SwapIfBodyWithSubsequence,
+            // move subsequent statements to if body
+            MoveSubsequenceToElseBody,
+            // invert and generete else
+            WithElseClause,
+            // invert and generate else, keep if-body empty
+            MoveIfBodyToElseClause,
+            // invert and copy the exit point statement
+            WithSubsequenceExitPoint,
+            // invert and generate return, break, continue
+            WithNearmostJump,
+            // just invert the condition
+            WithNegatedCondition,
+        }
+
+        protected abstract class Analyzer<TIfStatementSyntax> : IAnalyzer
+            where TIfStatementSyntax : SyntaxNode
+        {
+            public async Task ComputeRefactoringsAsync(CodeRefactoringContext context, SyntaxNode ifNode)
+            {
+                var ifStatement = (TIfStatementSyntax)ifNode;
+
+                var textSpan = context.Span;
+                var document = context.Document;
+                var cancellationToken = context.CancellationToken;
+
+                var headerSpan = GetHeaderSpan(ifStatement);
+                if (!headerSpan.IntersectsWith(textSpan))
+                {
+                    return;
+                }
+
+                if (ifStatement.OverlapsHiddenPosition(cancellationToken))
+                {
+                    return;
+                }
+
+                if (!CanInvert(ifStatement))
+                {
+                    return;
+                }
+
+                SyntaxNode subsequenceSingleExitPointOpt = null;
+
+                InvertIfStyle invertIfStyle;
+                if (IsElselessIfStatement(ifStatement))
+                {
+                    invertIfStyle = GetInvertIfStyle(
+                        ifStatement,
+                        await document.GetSemanticModelAsync().ConfigureAwait(false),
+                        ref subsequenceSingleExitPointOpt);
```

Comment:
this being ref feels hinky.  why not just have this method return both values?

Reply:
(bydesign)